### PR TITLE
feat(api): Implement stable pagination

### LINKED ISSUE #457: Provide stable sorting order in triggers API
Some automation tools require getting trigger information in bulk. For example, someone may want to ensure that there are no extra triggers that have a specific tag. Two potential problems arise with `GET triggers` API if there is a lot of triggers with a particular tag (order of tens of thousands):

1. If you request a page with size 10000, it will render several minutes and put a significant load on API service and Redis.
2. If you request several smaller pages one after another, you will encounter unstable sorting order that will make your task impossible to implement with 100% certainity.

We can either improve `GET triggers` rendering time to support large pages, or provide a stable sorting order to support sequential requests of smaller pages.

### PR DESCRIPTION
**Pull request by litleleprikon**
# PR Summary

Add ability to have stable pagination on GET /api/triggers/search.

Closes #457

### CODE REVIEW

File: database/redis/triggers_search_results.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+// GetTriggersSearchResults is a function that receives a saved pager from redis
+func (connector *DbConnector) GetTriggersSearchResults(searchResultsID string, page, size int64) ([]*moira.SearchResult, int64, error) {
+	c := connector.pool.Get()
+	defer c.Close()
+
+	var from, to int64 = 0, -1
+	if size > 0 {
+		from = page * size
+		to = from + size
```
Comment (litleleprikon) [inline database/redis/triggers_search_results.go]: Wrong to calculation. Should be 

```suggest
		to = from + size - 1
```

Comment (litleleprikon) [reply database/redis/triggers_search_results.go]: Fixed

File: database/redis/reply/search_result.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	Field string `json:"field"`
+	Value string `json:"value"`
+}
+
+type searchResultStorageElement struct {
+	ObjectID   string                          `json:"object_id"`
+	Highlights []searchHighlightStorageElement `json:"highlights"`
+}
+
+func toSearchResultStorageElement(searchResult moira.SearchResult) searchResultStorageElement {
```
Comment (beevee) [inline database/redis/reply/search_result.go]: I'm not sure, but maybe these functions are better written as methods of `searchResultStorageElement`?

`searchResultStorageElement.fromSearchResult` and `searchResultStorageElement.toSearchResult`?

File: database/redis/reply/search_result.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+		result.Highlights = append(result.Highlights, moira.SearchHighlight{
+			Field: highlight.Field,
+			Value: highlight.Value,
+		})
+	}
+	return result
+}
+
+// SearchResult is a function that converts redis reply to SearchResult.
+func SearchResult(rep interface{}, err error) (moira.SearchResult, error) {
```
Comment (beevee) [inline database/redis/reply/search_result.go]: I'm nervous that function name doesn't contain any verbs. SearchResult what?

File: database/redis/reply/search_result.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	err = json.Unmarshal(bytes, &storageElement)
+	if err != nil {
+		return searchResult, fmt.Errorf("failed to parse searchResult json %s: %w", string(bytes), err)
+	}
+	searchResult = toSearchResult(storageElement)
+	return searchResult, nil
+}
+
+// SearchResults is a function that converts redis reply to slice of SearchResults.
+func SearchResults(rep interface{}, repTotal interface{}, err error) ([]*moira.SearchResult, int64, error) {
```
Comment (beevee) [inline database/redis/reply/search_result.go]: Why return a slice of pointers to `SearchResult`? It's generally a bad idea: https://philpearl.github.io/post/bad_go_slice_of_pointers/

File: database/redis/triggers_search_results.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+import (
+	"fmt"
+
+	"github.com/gomodule/redigo/redis"
+	"github.com/moira-alert/moira"
+	"github.com/moira-alert/moira/database/redis/reply"
+)
+
+const triggersSearchResultsExpireTime = 1800
```
Comment (beevee) [inline database/redis/triggers_search_results.go]: 1800 what? Maybe rename constant to `triggersSearchResultsExpireTimeSeconds`. Or leave a comment.

File: api/controller/triggers.go [hunk 6]
```diff
@@ -57,10 +60,51 @@ func GetAllTriggers(database moira.Database) (*dto.TriggersList, *api.ErrorRespo
...
+			return nil, api.ErrorInternalServer(err)
+		}
+		pagerID = uuid4.String()
+		database.SaveTriggersSearchResults(pagerID, searchResults)
+	}
+
+	if createPager {
+		var from, to int64 = 0, int64(len(searchResults))
+		if size >= 0 {
+			from = int64(math.Min(float64(page*size), float64(len(searchResults))))
```
[the remark this answers is not in the document]
Comment (litleleprikon) [reply api/controller/triggers.go]: Would not cause the problem

File: database/redis/reply/search_result.go (hunk 2 above)
Comment (litleleprikon) [reply database/redis/reply/search_result.go]: I am doing this in a way this done in trigger.go do save consistency between a function signatures in this package.

https://github.com/moira-alert/moira/blob/a56078419152c5a9b85c2f0f7051442345ff1062/database/redis/reply/trigger.go#L33-L52

https://github.com/moira-alert/moira/blob/a56078419152c5a9b85c2f0f7051442345ff1062/database/redis/reply/trigger.go#L54-L73

File: database/redis/reply/search_result.go (hunk 3 above)
Comment (litleleprikon) [reply database/redis/reply/search_result.go]: Done in a way another functions named in this package 
https://github.com/moira-alert/moira/blob/a56078419152c5a9b85c2f0f7051442345ff1062/database/redis/reply/trigger.go#L87-L104

https://github.com/moira-alert/moira/blob/a56078419152c5a9b85c2f0f7051442345ff1062/database/redis/reply/notification.go#L13-L50

File: database/redis/reply/search_result.go (hunk 4 above)
Comment (litleleprikon) [reply database/redis/reply/search_result.go]: I do not like this way of working with slices of pointers to structures but reality is that Moira almost fully written in this manner. To use slice of structs here I need to rewrite an index and this is not covered by this task.

File: database/redis/triggers_search_results.go (hunk 5 above)
Comment (litleleprikon) [reply database/redis/triggers_search_results.go]: Fixed

File: go.mod [hunk 7]
```diff
@@ -10,6 +10,7 @@ require (
 	github.com/PagerDuty/go-pagerduty v0.0.0-20190503230806-cf1437c7c8d6
 	github.com/RoaringBitmap/roaring v0.4.21 // indirect
 	github.com/aclements/go-moremath v0.0.0-20190830160640-d16893ddf098 // indirect
+	github.com/alecthomas/jsonschema v0.0.0-20200217214135-7152f22193c9
```
[the remark this answers is not in the document]
Comment (litleleprikon) [reply go.mod]: Removed

**Pull request merged**